dango/lending/src/query.rs

```diff
@@ -64,3 +77,64 @@ pub fn query_debts(
         .take(limit as usize)
         .collect()
 }
+
+pub fn query_collateral_powers(
+    storage: &dyn Storage,
+) -> StdResult<BTreeMap<Denom, CollateralPower>> {
+    COLLATERAL_POWERS.load(storage)
+}
+
+// Mock oracle price for now.
+// TODO: Update once we finish oracle implementation.
+// Perhaps we want to make a "oracle querier" that has a cache? So that we don't
+// have to query the price for the same denom twice.
+pub fn query_oracle_price(_querier: &QuerierWrapper, _denom: &Denom) -> StdResult<Udec128> {
+    Ok(Udec128::ONE)
+}
+
+/// Calculates the health of a margin account.
+pub fn calculate_account_health(
+    querier: &QuerierWrapper,
+    margin_account: Addr,
+    debts: Coins,
+    collateral_powers: BTreeMap<Denom, CollateralPower>,
+) -> anyhow::Result<HealthResponse> {
+    // Calculate the total value of the debts.
+    let mut total_debt_value = Udec128::ZERO;
+    for coin in debts {
+        let price = query_oracle_price(querier, &coin.denom)?;
+        total_debt_value += coin.amount.checked_into_dec()? * price;
+    }
+
+    // Calculate the total value of the account's collateral adjusted for the collateral power.
+    let mut total_adjusted_collateral_value = Udec128::ZERO;
+    for (denom, power) in collateral_powers {
+        let price = query_oracle_price(querier, &denom)?;
+        let collateral_balance = querier.query_balance(margin_account, denom)?;
+        let collateral_value = collateral_balance.checked_into_dec()? * price;
+        total_adjusted_collateral_value += collateral_value * power.into_inner();
+    }
```

Comment:
There is an efficiency problem in this cycle. If the lending allows multiple collaterals, it is necessary to perform a query_balance and a query_price for each collateral. With time and the listing of new collaterals this part becomes very expensive. A solution would be to allow in the entry_point receive only the tokens managed by the lending pool and to save the various denominations. Then in calculate_account_health, perform a query on the balance of the tokens at the bank, if the quantity is 0 remove from the map managed by entry_point receive and go to the next, otherwise query the price and add the value to total_adjusted_collateral_value

Reply:
Hmm. @larry0x How many collateral tokens do we expect? Also how much in general do we care about such performance issues? I figured we're not in such a constrained environment as EVM and can easily handle iterating over a few dozen tokens without issue?

I also think it's a bit of a shame to restrict margin accounts only to hold those tokens which are listed as collateral tokens. Personally don't like making such restriction unless really necesssary. The more freedom people have in the ways they want to use the product the better imo.

@Rhaki you bring up a good point though that we don't need to query the price from the oracle if the collateral balance is zero. So will change the order here and continue in the loop if balance is zero.